src/appleseed/renderer/kernel/lighting/bdpt/bdptlightingengine.cpp

```diff
@@ -90,52 +134,324 @@ namespace
+            }
+            else
+            {
+                const BDPTVertex& light_vertex = light_vertices[s - 1];
+                const BDPTVertex& camera_vertex = camera_vertices[t - 2];
+
+                if (light_vertex.m_bsdf == nullptr || camera_vertex.m_bsdf == nullptr ||
+                    light_vertex.m_bsdf_data == nullptr || camera_vertex.m_bsdf == nullptr)
+                {
+                    return;
+                }
+
+                DirectShadingComponents camera_eval_bsdf;
+                camera_vertex.m_bsdf->evaluate(
+                    camera_vertex.m_bsdf_data,
+                    false,   // Adjoint
+                    false,
+                    static_cast<Vector3f>(camera_vertex.m_geometric_normal),
+                    camera_vertex.m_shading_basis,
+                    static_cast<Vector3f>(normalize(light_vertex.m_position - camera_vertex.m_position)),
+                    static_cast<Vector3f>(camera_vertex.m_dir_to_prev_vertex),
+                    ScatteringMode::All,
+                    camera_eval_bsdf);
+
+                DirectShadingComponents light_eval_bsdf;
+                light_vertex.m_bsdf->evaluate(
+                    light_vertex.m_bsdf_data,
+                    true,   // Adjoint
+                    false,
+                    static_cast<Vector3f>(light_vertex.m_geometric_normal),
+                    light_vertex.m_shading_basis,
+                    static_cast<Vector3f>(normalize(camera_vertex.m_position - light_vertex.m_position)),
+                    static_cast<Vector3f>(light_vertex.m_dir_to_prev_vertex),
+                    ScatteringMode::All,
+                    light_eval_bsdf);
+
+                Spectrum geometry = compute_geometry_term(shading_context, shading_point, camera_vertex, light_vertex);
+                result = geometry * camera_eval_bsdf.m_beauty * light_eval_bsdf.m_beauty * camera_vertex.m_beta * light_vertex.m_beta;
+            }
+
+            // check if connection is near black or not
+            bool nearBlack = true;
+            for (size_t i = 0; i < result.size(); i++)
+            {
+                // need nearBlack() function for spectrum
+                if (result[i] > 1e-9)
+                {
+                    nearBlack = false;
+                    break;
+                }
+            }
+
+            if (nearBlack) { return; }
+
+            float numerator = compute_path_density(light_vertices, camera_vertices, s, t, s, t);
+            float denominator = 0.0;
+            /// TODO:: unhandled case where denominator = 0 (specular surface).
+
+            // [p = 0, q = s + t], [p = 1, q = s + t - 1] ... [p = s + t - 2, q = 2]
+            for (size_t i = 0; i <= s + t - 2; i++)
+            {
+                denominator += compute_path_density(light_vertices, camera_vertices, s, t, i, s + t - i);
+            }
+
+            float miWeight = numerator / denominator;
```

Comment:
Variables should be spelled in snake case: `mi_weight`. Shouldn't it be `mis_weight` actually? What does "mi" stands for?